contrib/mesos/pkg/minion/server.go

```diff
@@ -243,6 +252,22 @@ func (ms *MinionServer) Run(hks hyperkube.Interface, _ []string) error {
 		log.Fatalf("No API client: %v", err)
 	}
 	ms.clientConfig = clientConfig
+	c := client.NewOrDie(clientConfig)
+
+	// create or update node object on apiserver. This must be done synchronously
+	// because the kubelet will refuse launching pods if the labels do not match.
+	labels := map[string]string{}
+	if ms.nodeLabelString != "" {
+		err = json.Unmarshal([]byte(ms.nodeLabelString), &labels)
+		if err != nil {
+			log.Fatalf("Invalid node-labels string: %v", err)
+		}
+	}
+	log.Infof("Creating or updating node %s with labels %v", ms.nodeHostName, labels)
+	_, err = node.CreateOrUpdate(c, ms.nodeHostName, labels)
```

Comment:
I don't see why creating the node object is not something the minion should do. In the sense of the recent renamings in k8s we could also rename `km minion` to `km node`.

With the node creation in the executor, we have two choices:
- the executor tells the minion that it may start the kubelet
- the executor blocks the kubelet launch somehow, by some kind of communication with the kubelet.

A completely different solution might be to let the kubelet wait for the creation of its node object (if node creation is disabled in the kubelet itself, I think there is a flag for that). This will involve a change in the kubelet and a polling. Not sure how acceptable this is for upstream.